backend-api/docs/pets/get_all.md

```diff
@@ -0,0 +1,83 @@
+# Get All Pets
+
+Get all pets registered in the country code provided in the headers field.
+
+**URL**: `/v1/pets`
+**Method**: `GET`
+
+**Headers**:
+```
+X-CountryIsoCode: "PT"
+```
+
+### Success Response
+
+**Code**: `200 OK`
+
+**Content Examples:**
+
+For a given `X-CountryIsoCode` header with value `PT` the database will answer with a list of `Pet` resources.
```

Comment:
couldn't this be an optional query string parameter which filters the results?

Reply:
Each country has it own method of generating Ids, taking that approach I also think that search should be the same. When you search for a pet, it should be segmented by country, that way we can segregate data easily.

Reply:
I agree with you if we have a search endpoint that could search everywhere and with a variety of fields. Example: country-code, pet-name, owner-name, healthstatus, etc... These are just examples.

Reply:
Then I would favor putting this country id in the url like `v1/countries/{country-code}/pets/`. I'm just not a big fan of using headers for data properties that actually belong to the data model.

Reply:
Personally I don't favor that approach. 

I think we should have them in a custom header or in the authorization token. Going with the auth approach implies having accounts for those who consume this API, which I am not against (personally I prefer this).

For example, a clinic or hospital that uses this API should be authenticated and authorized to use these endpoints. Being that a clinic or hospital must have a physical location in a country for example (this should be a constraint in order to avoid fake accounts). 

I think with this use case and possible approach we can solve our disparities, what do you think?

This may also solve the multi-tenancy use case.

Reply:
Using headers to manage state is fine imo (like auth) but I would strongly recommend decoupling auth from domain/data model parameters like the country code. In the future if you want to have users managing other countries you have to change how the API infers the country parameter via the auth context and possibly even change how auth is done. 

If you had the country parameter as a parameter (header or URL) it would be the responsibility of the caller to figure out which parameter makes sense in the context of the app and the backend api would just have to make sure that the user has access to that resource, e.g., obtaining a list of countries that the user has access to and checking if the requested resource is on a country of that list.

If it was something like "preferred culture" I would gladly have it on the auth context for the user. In this case it wouldn't be something that is part of the resource's data model necessarily and it's something that we can safely assume a default one (e.g., "en-US") in cases of no auth context).

Now back to the original dilemma, header vs URL... What's the pros of having it as a custom header instead of in the URL? One pro of it being in the URL is that it's safer when it comes to caching issues since clients can cache the GET request on the URL alone. Furthermore, I also think it's easier to understand the concept of the resource that way but this is personal preference so if you disagree we can leave this second point out of the discussion.

Reply:
I think this needs to be further discussed. I strongly believe we can reach a middle term on this theme. 

So, for now, I'll mark the header a TBD and we won't use it. If we need further developments in the future we should make them, it an evolutional project, it doesn't need to be perfect at the start.

So what do you think if we just do it, then do it right (when we have more info), then make it better (with even more info on our hands).

Reply:
After an offline discussion I think I understand a bit better the domain model we are going for and I believe we have other ways to obtain the country (via other linked resources like owner or similar) so we are going to remove the country parameter from the `pet` model. 

We also talked about developing a simple auth provider at some point in the future when the backend-api is a bit more developed and we have a better grasp of what we need.